Sources/NIOPosix/NonBlockingFileIO.swift

```diff
@@ -687,6 +687,70 @@ public struct NonBlockingFileIO: Sendable {
         }
     }
 
+    private func createDirectory0(_ path: String, mode: NIOPOSIXFileMode) throws {
+        // Fail fast if not a directory or file exists
+        do {
+            var s = stat()
+            try Posix.stat(pathname: path, outStat: &s)
+            if (S_IFMT & s.st_mode) == S_IFDIR {
+                return
+            }
+            throw IOError(errnoCode: ENOTDIR, reason: "Not a directory")
+        } catch {
+            // ignore error
+        }
+
+        // Slow path
+        var index = path.index(before: path.endIndex)
+        while index != path.startIndex && path[index] == "/" {
+            index = path.index(before: index)
+        }
+
+        while index != path.startIndex, path[index] != "/" {
+            index = path.index(before: index)
+        }
+
+        if index > path.startIndex {
+            try self.createDirectory0(String(path.prefix(upTo: index)), mode: mode)
+        }
+
+        do {
+            try Posix.mkdir(pathname: path, mode: mode)
+        } catch {
+            do {
+                var s = stat()
+                try Posix.lstat(pathname: path, outStat: &s)
+                if (S_IFMT & s.st_mode) == S_IFDIR {
+                    return
+                }
```

Comment:
It'd be good to explain what this block is trying to do.

Reply:
added comments